tests/Spec.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE FlexibleContexts #-}
+module Main where
+
+import Prelude as P
+import Conduit
+import Control.Monad (void, when, zipWithM_)
+import Test.Hspec
+import Data.Conduit.Tar
+import System.Directory
+import Data.ByteString as S
+import System.IO
+import System.FilePath
+import Control.Exception
+
+main :: IO ()
+main = do
+    let baseTmp = "tar-conduit-tests"
+    isStack <- doesDirectoryExist ".stack-work"
+    let testPaths =
+            ["src", "./tests", "README.md", "ChangeLog.md", "LICENSE"] ++
+            if isStack
+                then [".stack-work", "./sample"]
+                else []
+    hspec $ do
+        describe "tar/untar" $ do
+            let tarUntarContent dir =
+                    runConduitRes $
+                    yield dir .| void tarFilePath .| untar (const (foldC >>= yield)) .| foldC
+            before (collectContent "src") $ it "content" (tarUntarContent "src" `shouldReturn`)
+        describe "tar/untar/tar" $ do
+            around (withTempTarFiles baseTmp) $
+                it "structure" $ \(fpIn, hIn, outDir, fpOut) -> do
+                    writeTarball hIn testPaths
+                    hClose hIn
+                    extractTarball fpIn (Just outDir)
+                    curDir <- getCurrentDirectory
+                    finally
+                        (setCurrentDirectory outDir >> createTarball fpOut testPaths)
+                        (setCurrentDirectory curDir)
+                    tb1 <- readTarball fpIn
+                    tb2 <- readTarball fpOut
+                    P.length tb1 `shouldBe` P.length tb2
+                    zipWithM_ shouldBe (fmap fst tb2) (fmap fst tb1)
+                    zipWithM_ shouldBe (fmap snd tb2) (fmap snd tb1)
+
+withTempTarFiles :: FilePath -> ((FilePath, Handle, FilePath, FilePath) -> IO c) -> IO c
+withTempTarFiles base =
+    bracket
+        (do tmpDir <- getTemporaryDirectory
+            (fp1, h1) <- openBinaryTempFile tmpDir (addExtension base ".tar")
+            let outPath = dropExtension fp1 ++ ".out"
+            return (fp1, h1, outPath, addExtension outPath ".tar")
+        )
+        (\(fp, h, dirOut, fpOut) -> do
+             hClose h
+             removeFile fp
+             doesDirectoryExist dirOut >>= (`when` removeDirectoryRecursive dirOut)
```

Comment:
Should you just delete the entire `tmpDir` itself?

Reply:
I don't think it is a good idea delete it ;)
> On Unix, getTemporaryDirectory returns the value of the TMPDIR environment variable or "/tmp" if the variable isn't defined. ...

Reply:
You're right, I misread. I was thinking of this function https://www.stackage.org/haddock/lts-10.3/temporary-1.2.1.1/System-IO-Temp.html#v:withSystemTempDirectory